qualtran/bloqs/rotations/hamming_weight_phasing.py

```diff
@@ -86,3 +89,71 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
             (HammingWeightCompute(self.bitsize, adjoint=True), 1),
             (ZPowGate(exponent=self.exponent), self.bitsize.bit_length()),
         }
+
+
+@attrs.frozen
+class HammingWeightPhasingViaPhaseGradient(GateWithRegisters):
+    r"""Applies $Z^{\text{exponent}}$ to every qubit of an input register of size `bitsize`.
+
+    See docstring of `HammingWeightPhasing` for more details about how hamming weight phasing works.
+
+    In this variant of Hamming Weight Phasing, instead of directly synthesizing $O(n.bit_length())$
+    rotations on the hamming weight register, we synthesize the rotations via an addition into the
+    phase gradient register. See reference [1] for more details on this technique.
+
+    Note: For most reasonable values of `bitsize` and `eps`, the naive `HammingWeightPhasing` would
+    have better constant factors than `HammingWeightPhasingViaPhaseGradient`. This is because, in
+    general, the primary advantage of using phase gradient is to reduce the complexity from
+    $O(n * log(1/eps))$ to O(log(1/eps) ** 2) (the phase gradient register is of size
+    $O(log(1/eps))$ and a scaled addition into the target takes $(b_{grad} - 2)(log(1/eps) + 2)$).
+    Therefore, to apply $n$ individual rotations on a target register of size $n$, the complexity is
+    independent of $n$ and is essentially a constant (scales only with log(1/eps)).
+    However, for the actual constant values to be better, value of $n$ needs to be $> log(1/eps)$.
+    In the case of hamming weight phasing, $n$ corresponds to the hamming weight register which itself
+    is $log(\text{bitsize})$. Thus, as `eps` becomes smaller, the required value of $\text{bitsize}$,
+    for the phase gradient version to become more performant, becomes larger.
```

Comment:
Question: Is this specific to hamming weight phasing? In many of the chemistry algorithms rotations are always costed assuming a phase gradient gate (so like  (n - 2) Toffolis).

Reply:
The statement regarding tradeoff is valid in general but it becomes relevant mainly in the case of hamming weight phasing. Let's look at some numbers for more details -

- Let $eps \approx 10^{-12}$, $\log_2(\frac{1}{eps}) \approx 40$. 
- Let $n$ be the size of your system register s.t. $|\Psi_{n}\rangle$ is an n-qubit state we care about. 
$$|\Psi_{n}\rangle = \sum_{x=1}^{2^n}\alpha_{x}|x\rangle$$ 
- Suppose you want to apply a phase $e^{i \gamma x}$ to each state $|x\rangle$. 
- Naively, this would involve performing $n$ rotations, one on each qubit in the system register. Thus, the cost of the phasing operation would be $O(n \log(1/eps))$
- Using phase gradient trick, we do a scaled addition to a phase gradient register of size $b_\text{grad}$. One scaled addition reduces to $b_\text{gamma}$ normal additions, each of cost $b_\text{grad}$. Generally, both these variables are set $\approx O(\log(1/eps))$ and thus the total cost is $O(\log(1/eps)^ 2)$
- So, whenever the system register $n$ is greater than $b_\text{gamma}$, using phase gradient trick is better than naive approach. Assuming a system register of size > 40 for any beyond classical application is a reasonable assumption so for general chemistry algorithms it makes sense to ignore this tradeoff and always use phase gradient trick.
- However, for HammingWeightPhasing, for a system register of size $n$ the first reduction reduces the problem to applying only $\log(n)$ rotations on the hamming weight register. For $\log(n)$ to be greater than $\log(1/eps)$, $n$ should be $>1/eps$ -- this is a much more constrained scenario and one needs to be careful of the tradeoff when doing the costing.

Reply:
makes sense, thanks for the explanation